app/src/main/java/org/mozilla/focus/widget/ProgressBar.kt

```diff
@@ -0,0 +1,53 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.widget
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.LayoutInflater
+import android.view.View
+import android.arch.lifecycle.Observer
+import android.widget.LinearLayout
+import android.widget.TextView
+import kotlinx.android.synthetic.main.fragment_browser.view.*
+import org.mozilla.focus.R
+import android.view.animation.AnimationUtils
+import org.mozilla.focus.architecture.NonNullObserver
+import org.mozilla.focus.fragment.BrowserFragment
+import org.mozilla.focus.utils.GifImageView
+
+
+
+class ProgressBar @JvmOverloads constructor(
+        context: Context, attrs: AttributeSet? = null, defStyle: Int = 0)
+    : LinearLayout(context, attrs, defStyle) {
+
+    private var browserFrag: BrowserFragment? = null
+
+    fun setBrowserFragment(browserFragment: BrowserFragment) {
+        this.browserFrag = browserFragment
+        val textView = findViewById<TextView>(R.id.url)
+
+        browserFrag!!.session.url.observe(browserFrag!!, Observer { url -> textView.text = url })
+        browserFrag!!.session.loading.observe(browserFrag!!,object : NonNullObserver<Boolean>() {
+            public override fun onValueChanged(loading: Boolean) {
+                if (loading) {
+                    progressBar.visibility = View.VISIBLE
+                } else {
+                    val animFadeOut = AnimationUtils.loadAnimation(context, android.R.anim.fade_out);
+                    progressBar.setAnimation(animFadeOut)
+                    progressBar.visibility = View.INVISIBLE
+                }
+            }
+        })
+    }
+
+    init {
+        LayoutInflater.from(context)
+                .inflate(R.layout.progress_bar, this, true)
+        val gifImageView = findViewById<View>(R.id.gifImageView) as GifImageView
```

Comment:
nit: If you import the Kotlin android extensions, it has references to all the views so you don't need to do `findViewById` (i.e. this line can just be removed)